### ISSUE TITLE
Consider moving `Unapply` out of cats

### ISSUE THREAD

**Issue by kailuowang** [typelevel/cats#1607]
After #1583, cats no longer uses the `Unapply` machinery to overcome SI-2712. 
There are other projects using it, e.g. @wheaties's [AutoLifts](https://github.com/wheaties/AutoLifts) and my [henkan](https://github.com/kailuowang/henkan) We should consider moving `Unapply` to a separate project since it's no longer needed by cats and it is not a cohesive part of the what cats' offering, i.e. lib for pure functional programming.

**Comment by ceedubs** [comment]
There was a pre-SI-2712 fix issue similar to this: #574. I think that it's largely irrelevant now, though.

FWIW, I'm 👍 on removing `Unapply` from Cats. Since the SI-2712 fix, it's a lot less common to want it than previously, and I don't think it's very closely related to Cats.

**Comment by wheaties** [comment]
Once I get TwoTails to a point, I will happily remove Unapply from AutoLifts. Just not enough free time to prioritize.

**Comment by milessabin** [comment]
I think it would be good to remove it before Cats 1.0.

**Comment by kailuowang** [comment]
How about we move this to kittens? @milessabin

**Comment by milessabin** [comment]
Haven't we just removed unapply from Kittens though?

**Comment by kailuowang** [comment]
@milessabin yes we did. I am not sure if kittens is the right place for it either. I think the `Unapply` machinery could be helpful for scenarios where the left-to-right partial unification doesn't apply, but I don't know where it should go either. Maybe everyone just copy it to their own project since the usage will be so limited?

**Comment by johnynek** [comment]
What about cats-unapply with no dependency on cats but in the same package
name. Current users just add a dep to cats-unapply and no other change.
Probably the code never needs to change and we can delete it sometime later
if we really want to, but keeping it around probably does no harm.

**Comment by ceedubs** [comment]
@johnynek I could get behind a micro-library that just provides `Unapply` (though I personally don't have a use-case for it). But I really don't see any motivation to tie it to Cats. Cats wouldn't depend on it, and it wouldn't depend on Cats.

**Comment by kailuowang** [comment]
@ceedubs  that's a good point. Another thing that we could move to this micro library would be the `Trivial` class removed in this PR https://github.com/typelevel/cats/pull/1621. If we decide to create a new repo for them, shall we create it directly under typelevel github org, or shall we create under another account and go through the incubation process?

**Comment by johnynek** [comment]
See that's why I wouldn't do it. If others are using Unapply adding
cats-unapply is no big pain, and easy for us to do. If it becomes a bunch
of work it defeats the purpose for everyone involved (type renaming in any
files, for instance).

If we don't want a submodule for reasons, my second argument would be just
keep it but don't use it.

My third vote is just delete it.

**Comment by milessabin** [comment]
@kailuowang kittens already has `Trivial1` ... I'd be happy to see `Trivial` join it.
